### PR TITLE
fix: try/catch html parsing to allow non-html

### PR DESCRIPTION
**Pull request by AndrewLeedham**
fixes: #65 

- Split out `applyViteHtmlTransform` into its own file, so it was easier to test.
- Added optional chaining to `insertAdjacentHTML` in case HTML gets through without a body.
- Added a check for `.html` at the start of `applyViteHtmlTransform`, returning just the content if we are not processing HTML.
- Refactored the NPM scripts to not specify `/src` so singular tests can be run more easily from the command line with `npm t <test-file-pattern>`. Currently, if you were to do that it would run every test.
  - Hope you don't mind this one, it is not related to the functionality fix. More of quality of life DX thing. Happy to remove it if you want to keep this PR focused.

### CODE REVIEW

**Review by Holben888 — COMMENTED** [review]
Looks good! Maybe we could add a small unit test to `toHyddrationLoadersApplied.test.js` as well? Just an assertion that, if you pass in valid XML, you'll get out the same data untouched.

**Comment by AndrewLeedham** [comment]
@Holben888 didn't notice tests were set up, my bad. Think I saw the issue about adding unit tests and was thinking that meant there wasn't any.

I just had a go at writing them and realised this won't actually fix the issue. It seems that the error is coming from https://github.com/slinkity/slinkity/blob/9033abad531a0aa060df7b15183663862aa5d482/src/plugin/index.js#L63, when trying to add a style tag to XML, as there is no body. 

The bit of code with the try/catch around it won't actually be run because of this check: https://github.com/slinkity/slinkity/blob/main/src/plugin/index.js#L81 I believe.

**Comment by Holben888** [comment]
> I just had a go at writing them and realised this won't actually fix the issue. It seems that the error is coming from
> 
> https://github.com/slinkity/slinkity/blob/9033abad531a0aa060df7b15183663862aa5d482/src/plugin/index.js#L63
> 
Ah ha right you are! I wouldn't want to add a `try / catch` here since we're applying our Vite transform at the end. Maybe we could check that `querySelector('body')` is not null before appending styles? Just add a little option chain `?` before calling `insertAdjacentHtml`.

You can also check that your change is working by running Slinkity against your local project. Check [the contributing docs over here](https://github.com/slinkity/slinkity/blob/main/CONTRIBUTING.md#building-and-packing-the-framework) to set that up!

**Comment by AndrewLeedham** [comment]
@Holben888 I did think of doing a null assertion with optional chaining. But would it make sense to do the `.html` check at the top of `applyViteHtmlTransform` instead, since we have the `outputPath` in that function?

Actually, I suppose doing both would be the best bet?

**Comment by Holben888** [comment]
> @Holben888 I did think of doing a null assertion with optional chaining. But would it make sense to do the `.html` check at the top of `applyViteHtmlTransform` instead, since we have the `outputPath` in that function?

@AndrewLeedham So we could, but that would prevent [this line](https://github.com/slinkity/slinkity/blob/9033abad531a0aa060df7b15183663862aa5d482/src/plugin/index.js#L67) from running (which could mess with live reloading from what I understand). Curious if the optional chaining is enough to fix your issue?

**Comment by AndrewLeedham** [comment]
> > @Holben888 I did think of doing a null assertion with optional chaining. But would it make sense to do the `.html` check at the top of `applyViteHtmlTransform` instead, since we have the `outputPath` in that function?
> 
> @AndrewLeedham So we could, but that would prevent [this line](https://github.com/slinkity/slinkity/blob/9033abad531a0aa060df7b15183663862aa5d482/src/plugin/index.js#L67) from running (which could mess with live reloading from what I understand). Curious if the optional chaining is enough to fix your issue?

Good point I'll verify if the optional chaining alone fixes it. I did spend a bit of time breaking out `applyViteHtmlTransform` into its own file so I could test it. So wondering if it is worth add something like:
```javascript
  if (!outputPath.endsWith('.html')) {
    return environment === 'dev'
      ? viteSSR.server.transformIndexHtml(routePath, content)
      : content
  }
```
so we don't bother parsing it if we know it is not HTML?

**Comment by AndrewLeedham** [comment]
The optional chaining did not fix it:
```bash
`Error` was thrown:
[11ty]     Error: Unable to parse {"file":"/feed/feed.xml","line":1,"column":2}
    1  |  <?xml version="1.0" encoding="utf-8"?>
       |   ^
    2  |  <feed xmlns="http://www.w3.org/2005/Atom">
    3  |        <title>Your Blog Name</title>
        at traverseHtml (/Users/andrew/Documents/GitHub/Repro/eleventy-base-blog-slinkity/node_modules/vite/dist/node/chunks/dep-85dbaaa7.js:21240:15)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        at async devHtmlHook (/Users/andrew/Documents/GitHub/Repro/eleventy-base-blog-slinkity/node_modules/vite/dist/node/chunks/dep-85dbaaa7.js:56997:5)
        at async applyHtmlTransforms (/Users/andrew/Documents/GitHub/Repro/eleventy-base-blog-slinkity/node_modules/vite/dist/node/chunks/dep-85dbaaa7.js:21550:21)
        at async viteTransformMiddleware (/Users/andrew/Documents/GitHub/Repro/eleventy-base-blog-slinkity/node_modules/slinkity/lib/plugin/index.js:137:23)
 ```
However, neither did my proposed fix because that error is being thrown from `transformIndexHtml`.

**Comment by AndrewLeedham** [comment]
@Holben888 Apologies for the spam. But thinking about this more, I don't think we can have HMR on a non HTML page, as that requires injected scripts, which are not valid in XML or JSON. So perhaps the solution here is just returning the content early? (eleventy doesn't refresh the page for XML)

File: package.json [hunk 1]
```diff
@@ -87,7 +87,10 @@
     "watch": "babel src -d lib -w",
     "lint": "eslint \"src/**/*.{js,jsx,mjs}\"",
     "lint:fix": "eslint --fix \"src/**/*.{js,jsx,mjs}\"",
-    "test": "jest ./src",
-    "test:watch": "jest ./src --watch"
+    "test": "jest",
+    "test:watch": "jest --watch"
+  },
+  "jest": {
+    "roots": ["<rootDir>/src/"]
```
Comment (Holben888) [inline package.json:94]: Oh nice, never knew about this

**Review by Holben888 — APPROVED** [review]
Assuming everything works as before, this is a great refactor. Thanks for this!

**Comment by Holben888** [comment]
> @Holben888 Apologies for the spam. But thinking about this more, I don't think we can have HMR on a non HTML page, as that requires injected scripts, which are not valid in XML or JSON. So perhaps the solution here is just returning the content early? (eleventy doesn't refresh the page for XML)

Haha yep, I realized the same thing over dinner but you beat me to it. I agree we can return early by file extension.

**Comment by Holben888** [comment]
@AndrewLeedham Also feel free to add yourself as a contributor on our homepage! You can [follow @ajcwebdev's PR](https://github.com/slinkity/slinkity/pull/46) as a guide.

**Pull request merged**